pkg/common_controller/snapshot_controller.go

```diff
@@ -502,362 +645,270 @@ func (ctrl *csiSnapshotController) isVolumeBeingCreatedFromSnapshot(snapshot *cr
+			snapshotCopy.Status = &crdv1.VolumeSnapshotStatus{}
 		}
-		// Save failed, try again after a while.
-		klog.V(3).Infof("failed to save volume snapshot content %q for snapshot %q: %v", snapshotContent.Name, snapshotKey(snapshot), err)
-		time.Sleep(ctrl.createSnapshotContentInterval)
-	}
+		snapshotCopy.Status.BoundVolumeSnapshotContentName = &snapshotContent.Name
 
-	if err != nil {
-		// Save failed. Now we have a snapshot asset outside of Kubernetes,
-		// but we don't have appropriate volumesnapshot content object for it.
-		// Emit some event here and controller should try to create the content in next sync period.
-		strerr := fmt.Sprintf("Error creating volume snapshot content object for snapshot %s: %v.", snapshotKey(snapshot), err)
-		klog.Error(strerr)
-		ctrl.eventRecorder.Event(newSnapshot, v1.EventTypeWarning, "CreateSnapshotContentFailed", strerr)
-		return nil, newControllerUpdateError(snapshotKey(snapshot), err.Error())
+		updateSnapshot, err := ctrl.clientset.SnapshotV1beta1().VolumeSnapshots(snapshot.Namespace).Update(snapshotCopy)
+		if err != nil {
+			klog.Infof("bindVolumeSnapshotContentToVolumeSnapshot: Error binding VolumeSnapshot %s to volumeSnapshotContent [%s]. Error [%#v]", snapshot.Name, snapshotContent.Name, err)
+			return nil, newControllerUpdateError(utils.SnapshotKey(snapshot), err.Error())
+		}
+		snapshotCopy = updateSnapshot
+		_, err = ctrl.storeSnapshotUpdate(snapshotCopy)
+		if err != nil {
+			klog.Errorf("%v", err)
+		}
 	}
-	return newSnapshot, nil
-}
 
-// Delete a snapshot
-// 1. Find the SnapshotContent corresponding to Snapshot
-//   1a: Not found => finish (it's been deleted already)
-// 2. Ask the backend to remove the snapshot device
-// 3. Delete the SnapshotContent object
-// 4. Remove the Snapshot from store
-// 5. Finish
-func (ctrl *csiSnapshotController) deleteSnapshotContentOperation(content *crdv1.VolumeSnapshotContent) error {
-	klog.V(5).Infof("deleteSnapshotOperation [%s] started", content.Name)
-
-	// get secrets if VolumeSnapshotClass specifies it
-	var snapshotterCredentials map[string]string
-	/* TODO(@Xing-yang): secrete ref retrivial needs to be implemented here
-	  	snapshotClassName := content.Spec.VolumeSnapshotClassName
-		if snapshotClassName != nil {
-			if snapshotClass, err := ctrl.classLister.Get(*snapshotClassName); err == nil {
-				// Resolve snapshotting secret credentials.
-				// No VolumeSnapshot is provided when resolving delete secret names, since the VolumeSnapshot may or may not exist at delete time.
-				snapshotterSecretRef, err := getSecretReference(snapshotClass.Parameters, content.Name, nil)
-				if err != nil {
-					return err
-				}
-				snapshotterCredentials, err = getCredentials(ctrl.client, snapshotterSecretRef)
-				if err != nil {
-					return err
-				}
+	// update snapshot status
+	if snapshotContent.Status != nil {
+		// Update snapshot status with creationTime
+		for i := 0; i < ctrl.createSnapshotContentRetryCount; i++ {
```

Comment:
maybe this wrap this for loop into a small function? there are multiple places using the similar pattern.